### PR TITLE
Revise SDL-0282 Screen Manager Alert Manager

### PR DESCRIPTION
**Pull request by NicoleYarroch**
# Introduction
Update SDL-0282 Screen Manager Alert Manager to slightly modify the public APIs.

# Motivation
The audio data APIs should be updated for better usability. The current implementation assumes that an alert can only play one audio file or text-to-speech (TTS) chunk, however an alert can play as many sound files or TTS chunks as the developer wishes. Since most developers will only want to play one sound file or TTS chunk, the original convenience initializers were kept, however, now developers can choose to add additional TTS chunks and audio files if desired. In addition, the audio files and prompts are now stored in one list so the audio can be played in the desired order. It was also discovered that a `playTone` will only be played when the alert is displayed if an audio file or TTS chunk is set. The API was changed to only make it possible to set the `playTone` if an audio file or TTS chunk is set.

In addition, an `AlertCompletionListener` was added to the Screen Manager of Java-Suite library so the `tryAgainTime` can be returned to the developer if the alert fails to present. The iOS library can return the `tryAgainTime` in the existing `error` parameter.

# Proposed Solution
### iOS
```objc
/// SDLAudioData

/// Removed API
@property (nullable, copy, nonatomic, readonly) NSArray<SDLTTSChunk *> *prompt;
@property (nullable, copy, nonatomic, readonly) SDLFile *audioFile;

/// New API
@property (nullable, copy, nonatomic, readonly) NSArray<SDLTTSChunk *> *audioData;
- (instancetype)init NS_UNAVAILABLE;
- (void)addAudioFiles:(NSArray<SDLFile *> *)audioFiles;
- (void)addSpeechSynthesizerStrings:(NSArray<NSString *> *)spokenStrings;
- (void)addPhoneticSpeechSynthesizerStrings:(NSArray<NSString *> *)phoneticStrings phoneticType:(SDLSpeechCapabilities)phoneticType;
```

### Java
```java
/// AudioData

/// Removed API
private List<TTSChunk> prompt;
private SdlFile audioFile;

/// New API
private List<TTSChunk> audioData;
public void addAudioFiles(@NonNull List<SdlFile> audioFiles) {}
public void addSpeechSynthesizerStrings(@NonNull List<String> spokenString) {}
public void addPhoneticSpeechSynthesizerStrings(@NonNull List<String> phoneticString, @NonNull SpeechCapabilities phoneticType) {}
```

### Changes Made
1. Combined the `prompt` and `audioFile` properties into one array of `SDLTTSChunk`s called  `audioData`.
1. Made the default `init` unavailable so developers can not create an empty audio data object.
1. Added the `addAudioFiles:`, `addSpeechSynthesizerStrings:`, and `addPhoneticSpeechSynthesizerStrings:phoneticType:` methods to let developers add additional audio files or TTS chunks if desired. 

### iOS
```objc
/// SDLAlertAudioData

/// Removed API
- (instancetype)initWithTone;
- (instancetype)initWithAudioFile:(SDLFile *)audioFile playSystemTone:(BOOL)tone;
- (instancetype)initWithSpeechSynthesizerString:(NSString *)spokenString playSystemTone:(BOOL)tone;
- (instancetype)initWithPhoneticSpeechSynthesizerString:(NSString *)phoneticString phoneticType:(SDLSpeechCapabilities)phoneticType playSystemTone:(BOOL)tone;

/// New API
- (instancetype)init NS_UNAVAILABLE;
```

### Java
```java
/// AlertAudioData

/// Removed API
AlertAudioData(@NonNull Bool playTone)
AlertAudioData(@NonNull SdlFile audioFile, @NonNull Bool playTone)
AlertAudioData(@NonNull String spokenString, @NonNull Bool playTone)
AlertAudioData(@NonNull String phoneticString, @NonNull SpeechCapabilities phoneticType, @NonNull Bool playTone)

/// New API
AlertAudioData(@NonNull SdlFile audioFile)
AlertAudioData(@NonNull String spokenString)
AlertAudioData(@NonNull String phoneticString, @NonNull SpeechCapabilities phoneticType)
public AlertAudioData clone() // Creates a deep copy of the object
```

### Changes Made
1. Removed the convenience initializers for setting the `playTone` property.
1. Removed the convenience initializers for setting only the `playTone` property as at least one prompt or audio data file must be set in order for `playTone` to work. The default `init` was also made unavailable so developers can not create an empty audio data object.
1. Java-Suite only change: In order to implement copying, a public `clone` method had to be added to the `AlertAudioData` object. This ensures that the user can modify the `AlertAudioData` without changing previously sent `AlertAudioData`s. 

### iOS
```objc
/// SDLAlertView

/// Removed API
@property (nullable, copy, nonatomic) NSArray<SDLAlertAudioData *> *audio;

/// New API
@property (nullable, copy, nonatomic) SDLAlertAudioData *audio;
```

### Java
```java
/// AlertView

/// Removed API
private List<AlertAudioData> audio;

/// New API
private AlertAudioData audio;
public AlertView clone() // Creates a deep copy of the object
```

### Changes Made
1. An alert view now takes a single alert audio data object instead of a list of audio objects. This ensures that the `playTone` property is only set once instead of for each audio file or TTS chunk created.  
1. Java-Suite only change: In order to implement copying, a public `clone` method had to be added to the `AlertView` object. This ensures that the user can modify the `AlertView` without changing previously sent `AlertView`s.

## 4. Screen Manager Updates
```java
/// BaseScreenManager

/// Removed API
public void presentAlert(AlertView alert, CompletionListener listener)

/// New API
public void presentAlert(AlertView alert, AlertCompletionListener listener)
```

```java
/// New API
public interface AlertCompletionListener {
    /**
     * Returns whether an Alert operation was successful or not along with tryAgainTime
     * @param success - Boolean that is True if Operation was a success, False otherwise.
     * @param tryAgainTime - Amount of time (in seconds) that an app must wait before resending an alert.
     */
    void onComplete(boolean success, Integer tryAgainTime);
}
```
### Changes Made
1. Java-Suite only change: Created a new listener, `AlertCompletionListener`, so the `tryAgainTime` can be returned to the developer if an alert fails to present.
1. iOS only change: The documentation for the `presentAlert:withCompletionHandler` was updated let developers know that the `tryAgainTime` will returned in the `error` parameter of the `completionHandler`.  

## Updated Implementation Details 
1. Added additional implementation details about the behavior of the queue sending the alerts: 
> The `SDLAlertManager` sub-manager will use queues to manage alert related requests, similar to the implementation in the `SDLChoiceSetManager`. The queue is serial and if an alert is sent while another alert is currently presented, the newest alert will not be sent until the module dismisses the previous alert or `cancel` is called on the alert view by the developer on RPC 6.0+ systems.

# Potential downsides
None.

# Impact on existing code
None.

# Alternatives considered
No alternatives considered.

### CODE REVIEW

File: proposals/0282-screen-manager-alerts.md [hunk 1]
```diff
@@ -213,13 +202,13 @@ public class AlertView {
...
-        Builder setAudio(List<AlertAudioData> audio) { /* Imp */ }
+        Builder setAudio(AlertAudioData audio) { /* Imp */ }
         Builder setIcon(SdlArtwork icon) { /* Imp */ }
     }
 
-    // All vars have get / set as well to match iOS read / write
+    // All vars have getters and setters to match iOS read / write
 
-    public void cancel()
+    public void cancel() {}
```
Comment (JulianKast) [inline proposals/0282-screen-manager-alerts.md:211]:
```suggestion
    public void cancel() {}
    
    // Creates a deep copy of the object
    @Override
    public AlertView clone() 
```

**Review by JulianKast — COMMENTED** [review]
We also need to add `AlertCompletionListener` and change `BaseScreenManager` API to use it instead of a `CompletionListener`

```
public interface AlertCompletionListener {
    /**
     * Returns whether an Alert operation was successful or not along with tryAgainTime
     * @param success - Boolean that is True if Operation was a success, False otherwise.
     * @param tryAgainTime - Amount of time (in seconds) that an app must wait before resending an alert.
     */
    void onComplete(boolean success, Integer tryAgainTime);
}
```
BaseScreenManager

```
 public void presentAlert(AlertView alert, AlertCompletionListener listener)
```

**Review by renonick87 — APPROVED** [review]

File: proposals/0282-screen-manager-alerts.md [hunk 2]
```diff
@@ -245,7 +240,18 @@ And then the additions to the screen manager public API itself to present the al
 ##### Java
 ```java
 public class BaseScreenManager {
-    public void presentAlert(AlertView alert, CompletionListener listener)
+    // Everything already there
```
Comment (JulianKast) [inline proposals/0282-screen-manager-alerts.md:243]:
```suggestion
```

File: proposals/0282-screen-manager-alerts.md [hunk 3]
```diff
@@ -245,7 +240,18 @@ And then the additions to the screen manager public API itself to present the al
...
+    public void presentAlert(AlertView alert, AlertCompletionListener listener)
+}
+
+public interface AlertCompletionListener {
+    /**
+     * Returns whether an Alert operation was successful or not along with tryAgainTime
+     * @param success - Boolean that is True if Operation was a success, False otherwise.
+     * @param tryAgainTime - Amount of time (in seconds) that an app must wait before resending an alert.
+     */
+    void onComplete(boolean success, Integer tryAgainTime);
```
Comment (JulianKast) [inline proposals/0282-screen-manager-alerts.md:254]: Should we separate this into two blocks of code, I don't want to give the impression that the AlertCompletionListener is in the same file as BaseScreenManager, as it is its own file

**Review by JulianKast — CHANGES_REQUESTED** [review]

File: proposals/0282-screen-manager-alerts.md (hunk 3 above)
Comment (NicoleYarroch) [reply proposals/0282-screen-manager-alerts.md:254]: Separated into two blocks

**Review by JulianKast — APPROVED** [review]

File: proposals/0282-screen-manager-alerts.md [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Screen Manager Alert Manager
 
 * Proposal: [SDL-0282](0282-screen-manager-alerts.md)
-* Author: [Joel Fischer](https://github.com/joeljfischer)
+* Author: [Joel Fischer (https://github.com/joeljfischer), Nicole Yarroch (https://github.com/NicoleYarroch), Julian Kast (https://github.com/JulianKast)]
```
Comment (joeljfischer) [inline proposals/0282-screen-manager-alerts.md:4]:
```suggestion
* Author: [Joel Fischer](https://github.com/joeljfischer), [Nicole Yarroch](https://github.com/NicoleYarroch), [Julian Kast](https://github.com/JulianKast)
```

File: proposals/0282-screen-manager-alerts.md [hunk 5]
```diff
@@ -52,75 +46,69 @@ In order to support complicated `Alert`'s complicated audio processing and to si
...
 - (instancetype)initWithPhoneticSpeechSynthesizerString:(NSString *)phoneticString phoneticType:(SDLSpeechCapabilities)phoneticType;
-```
 
-```objc
-@interface SDLAlertAudioData: SDLAudioData
 /**
- Whether the alert tone should be played before the prompt (if any) is spoken. Defaults to NO.
- */
-@property (assign, nonatomic) BOOL playTone;
+ Add additional SDLFiles holding data or pointing to a file on the file system. When this object is passed to an `Alert` or `Speak`, the file will be uploaded if it is not already, then played if the system supports that feature.
```
Comment (joeljfischer) [inline proposals/0282-screen-manager-alerts.md:53]:
```suggestion
 Add additional `SDLFile`s holding data or pointing to a file on the file system. When this object is passed to an `Alert` or `Speak`, the file will be uploaded if it is not already, then played if the system supports that feature.
```

File: proposals/0282-screen-manager-alerts.md [hunk 6]
```diff
@@ -256,7 +262,7 @@ Due to the size of the iOS APIs and the similarity between the iOS, Java Suite a
 - The internal alert manager will observe the `AlertResponse` to know when the alert has finished presenting, and then call the `completionHandler`.
 - The internal alert manager will always send the alert, even if the system context is not MAIN. If the `AlertResponse` returns a failure to present, it will call the `completionHandler` with the error.
 - The developer will not be notified when the alert appears on the screen, assuming no error occurred – see alternative #1 for possible ways to do that.
-- The `SDLAlertManager` sub-manager will use queues to manage alert related requests, similar to how the `SDLChoiceSetManager` does.
+- The `SDLAlertManager` sub-manager will use queues to manage alert related requests, similar to the implementation in the `SDLChoiceSetManager`. The queue is serial and if an alert is sent while another alert is currently presented, the newest alert will not be sent until the module dismisses the previous alert.
```
Comment (joeljfischer) [inline proposals/0282-screen-manager-alerts.md:265]:
```suggestion
- The `SDLAlertManager` sub-manager will use queues to manage alert related requests, similar to the implementation in the `SDLChoiceSetManager`. The queue is serial and if an alert is sent while another alert is currently presented, the newest alert will not be sent until the module dismisses the previous alert or `cancel` is called on the alert view by the developer on RPC 6.0+ systems.
```

**Review by joeljfischer — CHANGES_REQUESTED** [review]

**Review by joeljfischer — APPROVED** [review]

**Comment by NicoleYarroch** [comment]
@pelsila Ready for review

**Review by pelsila — APPROVED** [review]

**Pull request merged**